46_revocation_severity_and_context.md

```diff
@@ -0,0 +1,301 @@
+is send and they are added to the database. Otherwise the event is ignored.
+
+In the verifier API for the agent a new field is exposed that contains the list
+of the revocation events, with their respected level and context.
+
+To the send revocation message two new field are added: `severity_level` and
+`context`. Such that the other agents can act accordingly.
+
+(To reduce bandwidth all the generated messages could also be bundled into one,
+but that would introduce more complexity of parsing the messages for the
+recipient.)
+
+If a agent gets removed from the verifier also all entries in
+`revocation_events` for the agent will be removed. (An option to clear the old
+events from a agent could be introduced, but a reset can already implemented by
+removing and then adding the agent to the verifier)
+
+### Events
+Events are tagged with an event id. Which has the following schema
+`component.[sub_component].event`.
+
+* `component`: Name of the part of Keylime where the event was generated.
+  * Current would be: `quote_validation`, `pcr_validation`, `measured_boot`,
+    `ima` and `internal`.
+* `sub_component`: If it is useful the separate a component into other sub
+  components. One example would be IMA checks.
+* `event`: The actual event itself.
+An example event id for a failing static PCR check would be `pcr_validation.pcr10`.
+
+In most cases the event ids static but for dynamic policies the `sub_component`
+and `event` can be generated automatically.
+
+The motivation behind that schema is to allow Keylime to adjust granularity of
+the events where needed and give the user an easy way to specify a severity
+level to one whole subset of Keylime.
+
+The severity level of an event is either 0 or if it matches one of the user
+rules the according to the configuration.
+
+Events can a specified context. Those can be either a static string or an JSON
+object that contains more information event. This is optional and it is assumed
+that two events from the same agent with the same event id have the same
+severity and are the same event.
+
+### User rules
+The user can specify severity and context for event ids. 
+One rule contains the following attributes:
+ * `event_id`: The event id to match or a regex rule for it.
+ * `severity_level`: The severity level that the matched events have.
+ * `context`: static string for the context of that event.
+
+To make the rules future proof with dynamic policies it is possible to specify a
+regular expression for that matching. This introduces more complexity on the
+parsing and matching side, but allows for flexible rules. Rules are parsed in a
+top down order and the first matching rule is used and are supplied as a JSON
+Array string to the verifier API (or the tenant).
+
+Rules are added to the agent similar on how it is done currently for `mb_refstate`.
+A new attribute `revocation_rules` is added to the agent to hold that information.
```

Comment:
This is a good idea and lets the policy be dynamic (rather than statically configured in keylime.conf) and lets whatever is adding the hosts for verification make that determination.

Reply:
That was my intention. I've added that now explicitly.